### DISCUSSION TITLE
"from pyraf.iraf import combine" fails?

### DISCUSSION THREAD

**Discussion by daniellekelley19** [iraf-community/iraf#331] [Q&A] [answered]
Hello everyone!

I am using jupyter notebook to write pyraf codes. Many tasks, like imslice, display and implot, could be soomthly invoked from the pyraf.iraf package just as illustrated in the tutorial. But this is not the case for the combine task and it returns an error indicating that this task is not included in pyraf.iraf package. Here, I know "help combine" could show the exact location of a given task in iraf syntax. But the fact is that lots of tasks have been reorganized and relocate in python syntax. So could anyone tell me how to use the combine task in python syntax? Thanks in advance!

Danielle

**Comment by olebole** [comment]
> But the fact is that lots of tasks have been reorganized and relocate in python syntax.

What do you mean here?  PyRAF is just another shell around IRAF (like cl/ecl) and so uses the task structure provided by IRAF. Therefore, the task is located at exactly the same place.

"help combine" shows on top where the task is located:
```
COMBINE (Aug96)             noao.imred.ccdred            COMBINE (Aug96)
```
So, the package is in noao.imred.ccdred.
If this doesn't help, please show what exactly you did, what you expected, and what you really got.

**Reply by daniellekelley19** [reply]
Hello Mr. Streicher!

Thanks very much for your reply! I reckon all tasks of the iraf sytax, whatever the subpackage they are belong to, are reflected in the pyraf.iraf package of python sytax. For exmaple, the display task is located in the images.tv package, so I can use "from _pyraf.iraf_ import display" rather than "from _pyraf.iraf.images.tv_ import display" command to load it to the python envornment. But combine task differs and "from pyraf.iraf import combine" can returns an error:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'combine' from 'pyraf.iraf' (/usr/lib/python3/dist-packages/pyraf/iraf.py)

It seems that this task merely appears in the iraf syntax or it is just located in a place other than pyraf.iraf. Could you please tell me how to use the combine task in python syntax and how to _search_ a task name among all the pyraf.Xs in python?

**Reply by olebole** [reply, marked as answer]
The following works for me:
```python
>>> from pyraf import iraf
>>> from iraf import noao, imred, ccdred
>>> iraf.combine()
List of images to combine: […]
```
The point is than one can only do top-level imports in iraf, not in the imported packages. Reason is the old-fashioned "import" behavior in PyRAF that reflects the IRAF package loading.

The difference to the other tasks you mentioned is that they are pre-loaded in your `login.cl` and therefore directly available in IRAF (and PyRAF).

One comment however: IRAF (and also PyRAF) is an old and outdated software. It is maintained for legacy and cultural reasons. Starting new projects with PyRAF or IRAF is not really a good idea; the time is usually better invested in using and improving the Astropy ecosystem.

**Reply by daniellekelley19** [reply]
Thanks very much for showing me how to locate the combine task and thanks for your suggestion as well!
